src/Microsoft.Extensions.OData.Migration/Middleware/ODataMigrationMiddleware.cs

```diff
@@ -1,175 +1,223 @@
+    using Microsoft.Data.OData.Query.SemanticAst;
+    using System;
+    using System.Collections.Specialized;
+    using System.IO;
+    using System.Linq;
+    using System.Net;
+    using System.Text.RegularExpressions;
+    using System.Threading.Tasks;
+    using System.Web;
+    using System.Xml;
+
+    /// <summary>
+    /// Translation Middleware currently converts V3 URI to V4 URI (future: converts query, request body as well)
+    /// </summary>
+    public class ODataMigrationMiddleware
+    {
+        private readonly RequestDelegate next;
+        private readonly Uri serviceRoot;
+        private readonly Data.Edm.IEdmModel v3Model;
+        private readonly Microsoft.OData.Edm.IEdmModel v4Model;
+
+        /// <summary>
+        /// Constructs an instance of TranslationMiddleware, requiring the root of the service, a V3 model instance and V4 model instance.
+        /// </summary>
+        /// <param name="next">Delegate required for middleware</param>
+        /// <param name="v3Edmx">V3 EDMX string representation of model</param>
+        /// <param name="v4Model">Instance of V4 EDM model</param>
+        public ODataMigrationMiddleware(RequestDelegate next,
+                                         string v3Edmx,
+                                         Microsoft.OData.Edm.IEdmModel v4Model)
+        {
+            ExceptionUtil.IfArgumentNullThrowException(v3Edmx, "v3Edmx", "V3 edmx cannot be empty");
+            ExceptionUtil.IfArgumentNullThrowException(v4Model, "v4Model", "V4 model cannot be null");
+
+            this.next = next;
+            this.serviceRoot = new Uri("http://localhost/"); // The actual service root doesn't matter; it is just needed as a parameter
+            this.v4Model = v4Model;
+
+            try {
+                using (XmlReader reader = XmlReader.Create(new StringReader(v3Edmx)))
+                {
+                    this.v3Model = Data.Edm.Csdl.EdmxReader.Parse(reader);
+                }
+            }
+            catch (Exception)
+            {
+                throw new ArgumentException("Unable to parse OData V3 Edmx; make sure your edmx is a valid OData contract.");
+            }
+        }
+
+        /// <summary>
+        /// Middleware method that conditionally modifies an HttpContext to be v4 compatible
+        /// </summary>
+        /// <param name="context">incoming HttpContext</param>
+        /// <returns></returns>
+        public async Task InvokeAsync(HttpContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(next));
+            }
+
+            // Fix Pascal case headers to lowercase, and default to JSON content type
+            ReplaceHeader(context.Request.Headers, "DataServiceVersion", "dataserviceversion");
+            ReplaceHeader(context.Request.Headers, "MaxDataServiceVersion", "maxdataserviceversion");
+            if (string.IsNullOrEmpty(context.Request.ContentType))
```

Comment:
We might need to check if the payload is XML. If so, throw an exception indicating that XML payload is not supported yet.